R/agglomerate.R

```diff
@@ -389,6 +418,67 @@ setMethod("agglomerateByVariable", signature = c(x = "SummarizedExperiment"),
     }
 )
 
+#' @rdname agglomerate-methods
+#' @export
+#' @importFrom S4Vectors DataFrame SimpleList
+#' @importFrom Matrix Matrix
+setMethod("agglomerateByModule", signature = c(x = "SummarizedExperiment"),
+    function(x, by, group, na.rm = FALSE){
+        # Check margin
+        by <- .check_MARGIN(by)
+        # Select side information based on margin
+        FUN <- switch(by, rowData, colData)
+        # Check group
+        if( !all(group %in% names(FUN(x))) ){
+            stop("'group' contained elements that did not match with any",
+                "column of ", as.character(substitute(rowData)), call. = FALSE)
+        }
+        # Extract module adjacency matrix as sparse array
+        modules <- as.matrix(FUN(x)[ , group, drop = FALSE])
+        # Check modules
+        is_logical <- is.logical(modules)
+        is_num <- all(modules == 0 | modules == 1)
+        is_na <- is.na(modules)
+        # Check validity of module type
+        if( !is_logical && !is_num ){
+            stop("'groups' variables are not binary.", call. = FALSE)
+        }
+        # Convert modules to sparse array after necessary checks
+        modules <- Matrix(modules, sparse = TRUE)
+        # Convert to numeric binary to Boolean adjacency matrix
+        if( is_num ){
+            modules <- modules != 0
+        }
+        # Replace NAs
+        if( any(is_na) ){
+            warning("NAs were found in 'groups' variables and were removed",
+                "before agglomerating the experiment.", call. = FALSE)
+            # Zero out NA modules
+            modules[is.na(modules)] <- FALSE
+        }
+        # Merge assays by module
+        assays <- mapply(.agglomerate_module_assay, assayNames(x), assays(x),
+            MoreArgs = list(by = by, modules = modules, na.rm = na.rm),
+            SIMPLIFY = FALSE)
```

Comment:
You could consider moving these to internal functions

```
assays <- .agglomerate_modules(x, modules)
x <- .constuct_tse_from_modules(x, assays)
```

I guess the general recommendation for length of function is 50 lines